### PR TITLE
Support for missing values using masked dataset

### PR DESCRIPTION
**Pull request by frazane**
## Type of changes

- [ ] Bug fix
- [x] New feature
- [ ] Documentation / docstrings
- [ ] Tests
- [ ] Other

## Checklist

- [x] I've formatted the new code by running `poetry run pre-commit run --all-files --show-diff-on-failure` before committing.
- [ ] I've added tests for new code.
- [ ] I've added docstrings for the new code.

## Description

This PR introduces support for missing data points (missing `y`s) in a way that is compatible with JAX jit-compilation (by keeping shapes static). This could be particularly useful in meta-learning settings (where each task has a variable number of available training points), or in multi-output GPs where only some of the outputs are missing (@ingmarschuster).

- new attribute `mask` added to the `Dataset`object
- added code to computations of `log_prob` and to posterior predictions, such that masked training inputs have no influence on the results

For now I only tested this on a simple exact GP with gaussian likelihood.

### CODE REVIEW

File: gpjax/dataset.py [hunk 1]
```diff
@@ -15,30 +15,55 @@
...
     X: Optional[Num[Array, "N D"]] = None
     y: Optional[Num[Array, "N Q"]] = None
+    mask: Optional[Union[Bool[Array, "N Q"], None]] = _Missing()
 
     def __post_init__(self) -> None:
         r"""Checks that the shapes of $`X`$ and $`y`$ are compatible."""
         _check_shape(self.X, self.y)
+        if isinstance(self.mask, _Missing):
+            if self.y is not None:
+                if jnp.any(mask := jnp.isnan(self.y)):
```
Comment (ingmarschuster) [inline gpjax/dataset.py:61]: I find
```
mask = jnp.isnan(self.y)
if jnp.any(mask)
```
easier to read than the := operator. this hold for later as well but is mostly a question of style.

File: gpjax/gaussian_distribution.py [hunk 2]
```diff
@@ -149,7 +152,9 @@ def entropy(self) -> ScalarFloat:
             self.event_shape[0] * (1.0 + jnp.log(2.0 * jnp.pi)) + self.scale.log_det()
         )
 
-    def log_prob(self, y: Float[Array, " N"]) -> ScalarFloat:
+    def log_prob(
```
Comment (ingmarschuster) [inline gpjax/gaussian_distribution.py:155]: It would be great to add a test just for the masked case in `test_gaussian_distribution.py`. This way if something is off in this core functionality, one instantly knows.

**Review by ingmarschuster — CHANGES_REQUESTED** [review]
Mostly needed is a test for masked log_prob. The := operator is just a question of style

File: gpjax/dataset.py (hunk 1 above)
Comment (frazane) [reply gpjax/dataset.py:61]: The walrus operator is still a controversial feature after all 😅 I agree it's not really needed in this case but later it makes for a cleaner code imo

File: gpjax/gaussian_distribution.py (hunk 2 above)
Comment (frazane) [reply gpjax/gaussian_distribution.py:155]: Yes, I have to add this and other tests.

File: gpjax/dataset.py (hunk 1 above)
Comment (thomaspinder) [reply gpjax/dataset.py:61]: I think I agree with Ingmar here.

File: gpjax/gaussian_distribution.py (hunk 2 above)
Comment (thomaspinder) [reply gpjax/gaussian_distribution.py:155]: +1

File: gpjax/dataset.py [hunk 3]
```diff
@@ -15,30 +15,55 @@
...
+    Bool,
+    Num,
+)
 from simple_pytree import Pytree
 
 from gpjax.typing import Array
 
 
+class _Missing:
+    """Sentinel class for not-yet-computed mask"""
```
Comment (thomaspinder) [inline gpjax/dataset.py:33]: Could you expand a little on this description please? To me, it doesn't fully explain the class' purpose.

Also, what is a sentinel class?

Comment (thomaspinder) [reply gpjax/dataset.py:33]: On reflection, why do we even need this class?

File: gpjax/dataset.py [hunk 4]
```diff
@@ -60,14 +85,20 @@ def __add__(self, other: "Dataset") -> "Dataset":
...
+        mask = None
 
         if self.X is not None and other.X is not None:
             X = jnp.concatenate((self.X, other.X))
 
         if self.y is not None and other.y is not None:
             y = jnp.concatenate((self.y, other.y))
 
-        return Dataset(X=X, y=y)
+        if (sm := self.mask is not None) | (om := other.mask is not None):
```
Comment (thomaspinder) [inline gpjax/dataset.py:96]: Stylistic, but elsewhere in the package we use `or` not `|`. Could we be consistent here please unless there's a performance benefit? Above discussion on `:=` holds true here too.

**Review by thomaspinder — CHANGES_REQUESTED** [review]

File: gpjax/dataset.py (hunk 3 above)
Comment (frazane) [reply gpjax/dataset.py:33]: The `Dataset` class holds either a boolean array or `None`, however at initialization `jnp.isnan` will be computed every time to ensure that there are no missing values in the data. This can be a lot of overhead especially when constructing batches (see `get_batch` in `fit.py`): if the original Dataset has `mask=None` there is no reason to do the computation for the subset. So the logic here is:
- if nothing is specified the default is this sentinel class, we compute the mask if needed
- if `None` is passed explicitly, we skip the computation and the mask is just `None`

In short, we need this class to provide a clear differentiation between an intentionally provided value of `None` and a default "unset" value for the mask attribute. 

In these situations the "sentinel object/value" pattern is useful, see https://python-patterns.guide/python/sentinel-object/

File: gpjax/dataset.py (hunk 4 above)
Comment (frazane) [reply gpjax/dataset.py:96]: "or" and "|" are not the same. If there are two conditions:
`cond1 or cond2`: cond2 does NOT evaluate if cond1 is already True
`cond1 | cond2`: both conditions are evaluated, and then True is returned is one of them is True (which is what we want here)

File: gpjax/dataset.py (hunk 3 above)
Comment (ingmarschuster) [reply gpjax/dataset.py:34]: This looks better to me than the solution I've always used (introduce an additional boolean variable `use_mask`). Maybe yet another would be to have a "Dataset" and a "MaskedDataset" class. I would be ok with the sentinel.

**Review by ingmarschuster — APPROVED** [review]
Sentinel would be fine to me. But see the comment about this above for other solution options I see

File: gpjax/dataset.py (hunk 3 above)
Comment (ingmarschuster) [reply gpjax/dataset.py:34]: another solution would be to use string constants, which is what some python libs do and you don't have to introduce a new class for it.

```
@dataclass
class Dataset:
    mask: Optional[Union[Bool[Array, "N Q"], str]] = "No Mask"
```

one hint: `Optional[...]` is a shorthand notation for `Union[..., None]` so your original type hint (` mask: Optional[Union[Bool[Array, "N Q"], None]] = _Missing()`) is doubling things.

**Comment by ingmarschuster** [comment]
I changed the sentinel class `_Missing` to a string literal alternative. I believe this serves the same purpose and might be more GPJax-style simple.
Accidentally pushed to the upstream branch. @frazane and @thomaspinder, what do you think? Shall we merge this version?
